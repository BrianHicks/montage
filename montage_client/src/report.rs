use serde::Serialize;

#[cynic::schema("montage")]
mod schema {}

#[derive(cynic::QueryVariables, Debug)]
pub struct ReportQueryVariables {
    pub start: DateTime,
    pub end: DateTime,
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "Query", variables = "ReportQueryVariables")]
pub struct ReportQuery {
    #[arguments(end: $end, start: $start)]
    pub report: Report,
}

#[derive(cynic::QueryFragment, Debug, Serialize)]
pub struct Report {
    pub start: DateTime,
    pub end: DateTime,
    pub totals: Totals,
    pub sessions: Vec<Session>,
}

#[derive(cynic::QueryFragment, Debug, Serialize)]
pub struct Totals {
    pub short_break: Duration,
    pub long_break: Duration,
    pub task: Duration,
    pub meeting: Duration,
    pub working: Duration,
    pub sessions_by_description: Vec<TotalByDescription>,
}

#[derive(cynic::QueryFragment, Debug, Serialize)]
pub struct TotalByDescription {
    pub description: String,
    pub kind: Kind,
    pub total: Duration,
}

#[derive(cynic::QueryFragment, Debug, Serialize)]
pub struct Session {
    pub description: String,
    pub actual_duration: Duration,
    pub kind: Kind,
    pub start_time: DateTime,
}

#[derive(cynic::Enum, Clone, Copy, Debug)]
pub enum Kind {
    Task,
    Break,
    Meeting,
}

type DateTime = chrono::DateTime<chrono::Local>;
cynic::impl_scalar!(DateTime, schema::DateTime);

type Duration = iso8601::Duration;
cynic::impl_scalar!(Duration, schema::Duration);

#[cfg(test)]
mod test {
    use super::*;
    use cynic::QueryBuilder;

    #[test]
    fn gql_output() {
        let operation = ReportQuery::build(ReportQueryVariables {
            start: chrono::Local::now(),
            end: chrono::Local::now(),
        });

        insta::assert_snapshot!(operation.query);
    }
}
