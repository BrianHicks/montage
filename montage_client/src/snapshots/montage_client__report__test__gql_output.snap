---
source: montage_client/src/report.rs
expression: operation.query
---
query ReportQuery($start: DateTime!, $end: DateTime!) {
  report(end: $end, start: $start) {
    start
    end
    totals {
      shortBreak
      longBreak
      task
      meeting
      working
      tasksByDescription {
        description
        total
      }
    }
    sessions {
      description
      actualDuration
      kind
      startTime
    }
  }
}


